packages/firestore/src/local/local_store.ts

```diff
@@ -642,6 +615,191 @@ export class LocalStore {
+    updatedKeys: DocumentKeySet,
+    documents: MaybeDocumentMap,
+    remoteVersion: SnapshotVersion | DocumentVersionMap
+  ): PersistencePromise<MaybeDocumentMap> {
+    const universalReadTime = remoteVersion instanceof SnapshotVersion;
+    return documentBuffer.getEntries(txn, updatedKeys).next(existingDocs => {
+      let changedDocs = maybeDocumentMap();
+      documents.forEach((key, doc) => {
+        const existingDoc = existingDocs.get(key);
+        let docReadVersion: SnapshotVersion | null = null;
+        if (universalReadTime) {
+          docReadVersion = remoteVersion as SnapshotVersion;
+        } else {
+          docReadVersion = (remoteVersion as DocumentVersionMap).get(key);
+        }
+        debugAssert(!!docReadVersion, 'Document read version must exist');
+
+        // Note: The order of the steps below is important, since we want
+        // to ensure that rejected limbo resolutions (which fabricate
+        // NoDocuments with SnapshotVersion.min()) never add documents to
+        // cache.
+        if (
+          doc instanceof NoDocument &&
+          doc.version.isEqual(SnapshotVersion.min())
+        ) {
+          // NoDocuments with SnapshotVersion.min() are used in manufactured
+          // events. We remove these documents from cache since we lost
+          // access.
+          documentBuffer.removeEntry(key, docReadVersion!);
+          changedDocs = changedDocs.insert(key, doc);
+        } else if (
+          existingDoc == null ||
+          doc.version.compareTo(existingDoc.version) > 0 ||
+          (doc.version.compareTo(existingDoc.version) === 0 &&
+            existingDoc.hasPendingWrites)
+        ) {
+          debugAssert(
+            !SnapshotVersion.min().isEqual(docReadVersion!),
+            'Cannot add a document when the remote version is zero'
+          );
+          documentBuffer.addEntry(doc, docReadVersion!);
+          changedDocs = changedDocs.insert(key, doc);
+        } else {
+          logDebug(
+            LOG_TAG,
+            'Ignoring outdated watch update for ',
+            key,
+            '. Current version:',
+            existingDoc.version,
+            ' Watch version:',
+            doc.version
+          );
+        }
+      });
+      return changedDocs;
+    });
+  }
+
+  /**
+   * Returns a promise of a boolean to indicate if the given bundle has already
+   * been loaded and the create time is newer than the current loading bundle.
+   */
+  isNewerBundleLoaded(
+    bundleMetadata: bundleProto.BundleMetadata
+  ): Promise<boolean> {
+    const currentReadTime = this.bundleConverter.toSnapshotVersion(
+      bundleMetadata.createTime!
+    );
+    return this.persistence
+      .runTransaction('isNewerBundleLoaded', 'readonly', transaction => {
+        return this.bundleCache.getBundle(transaction, bundleMetadata.id!);
+      })
+      .then(cached => {
+        return !!cached && cached!.createTime!.compareTo(currentReadTime) > 0;
```

Comment:
You only need to do the `!!` if you are assigning to a boolean. If statements except "truthy" values directly, so `!!cached` is the same as `cached`. 

You can also drop the bang on `cached!`.

Reply:
Here `!!` is used to remove the `| undefined` part of the return type, without it the return type mismatches.